lib/charms/loki_k8s/v0/loki_push_api.py

```diff
@@ -1282,31 +1250,35 @@ def __init__(
         self.framework.observe(
             self._charm.on.log_proxy_relation_departed, self._on_log_proxy_relation_departed
         )
+        self.framework.observe(
+            getattr(self._charm.on, "{}_pebble_ready".format(self._container_name)),
+            self._on_pebble_ready,
+        )
+
+    def _on_pebble_ready(self, event):
+        """Event handler for `pebble_ready`."""
+        if self.model.relations[self._relation_name] and not self._is_promtail_installed():
+            self._setup_promtail()
 
     def _on_log_proxy_relation_created(self, event):
         """Event handler for the `log_proxy_relation_created`."""
-        self._create_directories()
-        self._container.push(
-            WORKLOAD_CONFIG_PATH, yaml.safe_dump(self._initial_config), make_dirs=True
-        )
+        if not self._container.can_connect():
+            return
+        if not self._is_promtail_installed():
+            self._setup_promtail()
```

Comment:
The can_connect() guard is needed for code outside of _setup_promtail as well so it can't be moved inside. I think _is_installed probably can be though.

Reply:
Done for _is_promtail_installed.

Reply:
This actually broke things and I'm reverting it as I think it is not the behavior that would be expected from such a method and could break things.